### PR TITLE
Logs Panel: Table UI - Pull logs table into dashboard panel

### PR DESCRIPTION
**Pull request by gtk-grafana**
Part of https://github.com/grafana/grafana/issues/77469

**What is this feature?**
Allowing users to add a logs table from explore to a new or existing dashboard.
Uses the information in the explore url to pass transformations into table panel which will result in same view that user has configured in explore.

**Why do we need this feature?**
Users expect that they can create a table in explore and add it as a dashboard. This is a primary use-case for sharing results of logs investigation.

Depends on https://github.com/grafana/grafana/pull/77797 transform change and [base branch refactor](https://github.com/grafana/grafana/pull/78070)

### CODE REVIEW

File: public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts [hunk 1]
```diff
@@ -29,13 +31,42 @@ function createDashboard(): DashboardDTO {
...
+    });
+  }
+
   const panel = {
     targets: options.queries,
     type: panelType,
     title: 'New Panel',
     gridPos: { x: 0, y: 0, w: 12, h: 8 },
     datasource: options.datasource,
+    fieldConfig: { defaults: {}, overrides: [] },
```
Comment (gtk-grafana) [inline public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts:68]: remove

File: public/app/features/explore/Logs/LogsTableWrap.tsx [hunk 2]
```diff
@@ -249,6 +249,8 @@ export function LogsTableWrap(props: Props) {
       setFilteredColumnsWithMeta(pendingFilteredLabelState);
     }
 
+    const logsFrame = parseLogsFrame(dataFrame);
```
Comment (gtk-grafana) [inline public/app/features/explore/Logs/LogsTableWrap.tsx:252]: parsing the log frame in explore? there has to be a better way

File: public/app/features/explore/Logs/LogsTableNavColumn.tsx [hunk 3]
```diff
@@ -72,6 +72,16 @@ function sortLabels(labels: Record<string, fieldNameMeta>) {
       }
     }
 
+    if (labels[b].active && labels[a].active) {
```
Comment (gtk-grafana) [inline public/app/features/explore/Logs/LogsTableNavColumn.tsx:75]: this all gets cleaned up in https://github.com/grafana/grafana/pull/78150

**Comment by gtk-grafana** [comment]
Needs tests!

File: e2e/various-suite/loki-table-explore-to-dash.spec.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+      frames: [
+        {
+          schema: {
+            refId: 'A',
+            meta: {
+              typeVersion: [0, 0],
+              custom: {
+                frameType: 'LabeledTimeValues',
+              },
+              stats: [
```
Comment (gtk-grafana) [inline e2e/various-suite/loki-table-explore-to-dash.spec.ts:29]: can we remove some of this?

**Review by mckn — COMMENTED** [review]
Code looks good to me but I leave it up to the @grafana/explore-squad to approve this PR.

**Comment by svennergr** [comment]
Hm, it didn't really work with a Loki query for me:

Explore: 
image

Dashboard will just remove all fields:
image

**Comment by gtk-grafana** [comment]
@svennergr should be fixed now, needed to clean up the state when the Logs component is unmounted.

File: public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts [hunk 5]
```diff
@@ -28,14 +30,77 @@ function createDashboard(): DashboardDTO {
...
+ * Returns the transformations to apply to the panel based on the panel type
+ * @param panelType
+ * @param options
+ */
+function getExploreToPanelTransformations(panelType: string, options: AddPanelToDashboardOptions) {
+  let transformations: DataTransformerConfig[] = [];
+  transformations.push(...getLogsTableTransformations(panelType, options));
+
+  return transformations;
+}
```
Comment (Elfo404) [inline public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts:80]: this seems to just proxy the call to `getLogsTableTransformations`, can we just use `getLogsTableTransformations` on line 103?

```suggestion
```

File: public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts [hunk 6]
```diff
@@ -28,14 +30,77 @@ function createDashboard(): DashboardDTO {
...
+ * Filters the query for logs table, if the user has a query selected in explore, only return that one
+ * The logs table uses organize fields transformation, which only supports a single dataframe
+ * Also, the UI only displays one dataframe/query at a time, so this should result in the expected behavior.
+ * @param options
+ */
+function getExploreToPanelQueries(options: AddPanelToDashboardOptions) {
+  return options?.panelState?.logs?.refId
+    ? [options.queries.find((q) => q.refId === options?.panelState?.logs?.refId)]
+    : options.queries;
+}
```
Comment (Elfo404) [inline public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts:92]: I'm not sure about this, if the user has a a metric query (or any other result type) and a logs query, the metrics query is completely ignored. i think the source of the issue here is that e2d adds everything to a single panel, and this doesn't really scale well in this scenario

Comment (gtk-grafana) [reply public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts:80]: Yeah, so I guess we can either pull everything into the new panel, just pull the first query into the panel, or create a panel for each.

But on main I'm seeing this same issue, when creating multiple queries, the query type of the first query defines the panel type, and all other queries in explore are dumped into the same dashboard panel.

I would argue that this is a pre-existing issue that isn't introduced in this PR? What do you think @Elfo404?

Comment (Elfo404) [reply public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts:80]: the problem here though is that we are removing queries, while on main what happens is that all the queries are pulled into the panel, and the panel itself is - as you said - picked from the available ones based on the result.

we probably need to do something else in explore, and rethink a bit how the "add to dashboard" works, possibly being a bit smarter  about the number of panels/queries.

cc @grafana/explore-squad

Comment (Elfo404) [reply public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts:80]: i'd say that would be the best thing to do for now as it is the documented behaviour

Comment (gtk-grafana) [reply public/app/features/explore/extensions/AddToDashboard/addToDashboard.ts]: Should be pushed up now

**Comment by Elfo404** [comment]
Explore looks good! i'll let the logs squad approve

File: e2e/various-suite/loki-table-explore-to-dash.spec.ts
Comment (svennergr) [inline e2e/various-suite/loki-table-explore-to-dash.spec.ts:1]: Nice!!

File: public/app/features/explore/Logs/Logs.tsx [hunk 7]
```diff
@@ -183,6 +192,7 @@ class UnthemedLogs extends PureComponent<Props, State> {
           ...state.panelsState.logs,
           columns: logsPanelState.columns ?? this.props.panelState?.logs?.columns,
           visualisationType: logsPanelState.visualisationType ?? this.state.visualisationType,
+          labelName: logsPanelState.labelName,
```
Comment (svennergr) [inline public/app/features/explore/Logs/Logs.tsx:195]: Could we refactor `labelName` to `labelFieldName` please? I think that name makes it a lot clearer what this prop is about.

**Review by svennergr — COMMENTED** [review]
I think we need to add a transformation to select the dataframe that was previously selected in the Explore table vis?

image

**Comment by gtk-grafana** [comment]
> I think we need to add a transformation to select the dataframe that was previously selected in the Explore table vis?
> 
See the conversation with Gio here: https://github.com/grafana/grafana/pull/77757#discussion_r1405971715, the current functionality in explore is to dump everything into the panel, and we didn't want to change that in this PR.

**Review by svennergr — APPROVED** [review]
Nice!

**Pull request merged**